controller/api/destination/watcher/endpoints_watcher.go

```diff
@@ -194,18 +195,22 @@ func (ew *EndpointsWatcher) addService(obj interface{}) {
 }
 
 func (ew *EndpointsWatcher) deleteService(obj interface{}) {
-	service := obj.(*corev1.Service)
-	if service.Namespace == kubeSystem {
-		return
-	}
-	id := ServiceID{
-		Namespace: service.Namespace,
-		Name:      service.Name,
-	}
+	if serviceObj, err := extractDeletedObject(obj, reflect.TypeOf(&corev1.Service{})); err == nil {
```

Comment:
I see that you're trying to factor the common code for handling the final state unknown into a common function.  Unfortunately, this introduces unnecessary use of "reflect" because Go doesn't support generics.  I'd recommend following the pattern of the k8s libraries (https://github.com/kubernetes/kubernetes/blob/0b2636a7e705bdf7c911e12d2e8846b9e6f88704/pkg/controller/replicaset/replica_set.go#L330-L342) and check the type against the final state unknown tombstone in each delete handler.

Yes, this is a lot more repeated code.  It is the Go way.  😞

Reply:
I see, but what is the problem with using the reflect package in this particular case ?

Reply:
I try to avoid reflection whenever reasonable.  According to the [Go proverbs](https://go-proverbs.github.io/):

>Clear is better than clever.
>Reflection is never clear.

Reply:
I have to agree, reflection is hacky. Will inline the checks and remove the tests as well. Not sure I like the Go way :(

Reply:
Took another look at this and with the risk of starting an argument I dont want to be a part of, here are my top reasons to still opt out for the reflection here: 

1. We seem to be using reflection in various places of the code (of course if its so bad the fact that we have done doesn't mean we should be doing it again).
2. If we are to do that we will most likely loose the ability to have meaningful tests on that logic.
3. Its hard for me to explain why the k8s codebase uses reflection in quite some spots as well but opted for code repetition in these particular places.

Reply:
Can you elaborate on point 2?  If we don't have a `extractDeletedObject` function then we cannot test it --- but we should still be able to test the individual delete handlers.

Reply:
Ok, sure we can test the individual handlers. Its just going to be a such a non DRY piece of code... Anyway, your call. I am fine with whatever approach

Reply:
Writing Go code is like trying to give my kids a bath.  I'd prefer to stay DRY but it never seems to work out that way...

Reply:
Hard to disagree :)

Reply:
@adleong That's it. We are drenched now. Feel free to take a second look :)